linalg/cudss.hpp

```diff
@@ -0,0 +1,181 @@
+
+   /**
+    * @brief Set the flag controlling reuse of the symbolic factorization
+    * for multiple operators
+    *
+    * @param reuse Flag to reuse symbolic factorization
+    *
+    * @note This method has to be called before repeated calls to SetOperator
+    */
+   void SetReorderingReuse(bool reuse);
+
+   /**
+    * @brief Set the flag controlling sort the rows of CSR matrix
+    *
+    * @param sort_row_ Flag to sort the row of CSR matrix
+    *
+    * @note This method has to be called before repeated calls to SetOperator
+    */
+   void SetMatrixSortRow(bool sort_row_);
+
+   void SetOperator(const Operator &op) override;
+   /**
+    * @brief Solve $ y = Op^{-1} x $
+    *
+    * @param x RHS vector
+    * @param y Solution vector
+    */
+   void Mult(const Vector &x, Vector &y) const override;
+
+   /**
+    * @brief Solve $ Y_i = Op^{-1} X_i $
+    *
+    * @param X Array of RHS vectors
+    * @param Y Array of Solution vectors
+    */
+   void ArrayMult(const Array<const Vector *> &X, Array<Vector *> &Y) const;
+
+   ~CuDSSSolver();
+
+private:
+   // MPI_Comm
+   MPI_Comm *mpi_comm = NULL;
+   // MPI rank
+   int myid;
+
+   // Parameter controlling whether or not to reuse the symbolic factorization
+   // for multiple calls to SetOperator
+   bool reorder_reuse = false;
+
+   // Parameter controlling whether or not to sort the row of CSR matrix
+   bool sort_row = false;
+
+   // Matrix type
+   MatType mtype = MatType::NONSYMMETRIC;
+   // Parameter controlling the matrix type
+   cudssMatrixType_t mat_type = CUDSS_MTYPE_GENERAL;
+
+   mutable int nrhs = 0; // the number of the RHSs
+   int n_global = 0;     // global number of rows
+
+   int nnz = 0; // the number of non zeros
+   int row_start = 0;  // the first row index in CSR matrix operator
+   int row_end = 0;   // the end row index in CSR matrix operator
+   int n_loc = 0;    // the number of the rows in CSR matrix operator
+
+   // copy and keep the I and J arrays in device memory when skipping analysis
+   // phase
+   int *csr_offsets_d = NULL; // copy and keep I in device
+   int *csr_columns_d = NULL; // copy and keep J in device
+
+   // cuDSS object specifies available matrix types for sparse matrices
+   cudssMatrixViewType_t mview = CUDSS_MVIEW_FULL;
+
+   // cuDSS objects storage for sparse matrix Ac, RHS yc and solution xc
+   std::unique_ptr<cudssMatrix_t> Ac;
+   mutable cudssMatrix_t xc, yc;
+
+   // cuDSS object holds the cuDSS library context
+   mutable cudssHandle_t handle;
+   // cuDSS object stores configuration settings for the solver
+   mutable cudssConfig_t solverConfig;
+   // cuDSS object holds internal data
+   mutable cudssData_t solverData;
+
+   /// Method for initialization
+   void Init(MPI_Comm comm_);
+
+   /// Method for the initialization of cuDSS Handle
+   void InitHandle();
+
+   /// Method for configuring storage for distributed/centralized RHS and
+   /// solution
+   void InitRhsSol(int nrhs_) const;
```

Comment:
In MFEM, usually written 'RHS'

```suggestion
   void InitRHSSol(int nrhs_) const;
```

Maybe a clearer name would be `SetNumRHS` or similar?